compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DocumentationAnalyzer.java

```diff
@@ -0,0 +1,256 @@
+            validateParameters(typeDefinition, fields, DiagnosticCode.UNDOCUMENTED_FIELD,
+                    DiagnosticCode.NO_SUCH_DOCUMENTABLE_FIELD, DiagnosticCode.FIELD_ALREADY_DOCUMENTED);
+            validateReturnParameter(typeDefinition, null, false);
+
+            ((BLangObjectTypeNode) typeDefinition.getTypeNode()).getFunctions().forEach(f -> analyzeNode(f, env));
+        } else if (typeDefinition.typeNode.getKind() == NodeKind.RECORD_TYPE) {
+            List<? extends VariableNode> fields = ((BLangRecordTypeNode) typeNode).getFields();
+            validateParameters(typeDefinition, fields, DiagnosticCode.UNDOCUMENTED_FIELD,
+                    DiagnosticCode.NO_SUCH_DOCUMENTABLE_FIELD, DiagnosticCode.FIELD_ALREADY_DOCUMENTED);
+            validateReturnParameter(typeDefinition, null, false);
+        }
+    }
+
+    @Override
+    public void visit(BLangResource resourceNode) {
+        validateParameters(resourceNode, resourceNode.getParameters(), DiagnosticCode.UNDOCUMENTED_PARAMETER,
+                DiagnosticCode.NO_SUCH_DOCUMENTABLE_PARAMETER, DiagnosticCode.PARAMETER_ALREADY_DOCUMENTED);
+
+        validateReturnParameter(resourceNode, null, false);
+    }
+
+    private void validateParameters(DocumentableNode documentableNode, List<? extends VariableNode> actualParameters,
+                                    DiagnosticCode undocumentedParameter, DiagnosticCode noSuchParameter,
+                                    DiagnosticCode parameterAlreadyDefined) {
+        BLangMarkdownDocumentation documentation = documentableNode.getMarkdownDocumentationAttachment();
+        if (documentation == null) {
+            return;
+        }
+
+        Map<String, BLangMarkdownParameterDocumentation> documentedParameterMap = new HashMap<>();
+        documentation.parameters.forEach(parameter -> {
+            String parameterName = parameter.getParameterName().getValue();
+            // Check for duplicate documentations.
+            if (documentedParameterMap.containsKey(parameterName)) {
+                dlog.warning(parameter.pos, parameterAlreadyDefined, parameterName);
+            } else {
+                documentedParameterMap.put(parameterName, parameter);
+            }
+        });
+
+        Map<String, VariableNode> actualParameterMap = new HashMap<>();
+        actualParameters.forEach(parameter -> {
+            String name = parameter.getName().getValue();
+            actualParameterMap.put(name, parameter);
+            // Set the symbol in the documentation node.
+            BLangMarkdownParameterDocumentation param = documentation.getParameterDocumentations().get(name);
+            if (param != null) {
+                param.setSymbol(((BLangVariable) parameter).symbol);
+            }
+        });
+
+        // Temporary list to hold matching parameter names.
+        List<String> matchedParameters = new LinkedList<>();
```

Comment:
Can't we reduced these three loops in to one? :)

Reply:
No, it will then throw concurrent modification exception.